alerta/auth/decorators.py

```diff
@@ -0,0 +1,102 @@
+import base64
+import re
+from flask import request, g, current_app
+from functools import wraps
+from jwt import DecodeError, ExpiredSignature, InvalidAudience
+
+from alerta.auth.utils import is_authorized, get_customers
+from alerta.exceptions import ApiError
+from alerta.models.key import ApiKey
+from alerta.models.permission import Permission
+from alerta.models.token import Jwt
+from alerta.models.user import User
+
+
+def permission(scope):
+    def decorated(f):
+        @wraps(f)
+        def wrapped(*args, **kwargs):
+
+            auth_header = request.headers.get('Authorization', '')
+
+            # API Key
+            m = re.match(r'Key (\S+)', auth_header)
+            param = m.group(1) if m else request.args.get('api-key', None)
+
+            if param:
+                key = ApiKey.verify_key(param)
+                if not key:
+                    raise ApiError("API key parameter '%s' is invalid" % param, 401)
+                g.user = key.user
+                g.customers = [key.customer] if key.customer else []
+                g.scopes = key.scopes
+
+                if not Permission.is_in_scope(scope, g.scopes):
+                    raise ApiError('Missing required scope: %s' % scope, 403)
+                else:
+                    return f(*args, **kwargs)
+
+            # Bearer Token
+            m = re.match(r'Bearer (\S+)', auth_header)
+            token = m.group(1) if m else None
+
+            if token:
+                try:
+                    jwt = Jwt.parse(token)
+                except DecodeError:
+                    raise ApiError('Token is invalid', 401)
+                except ExpiredSignature:
+                    raise ApiError('Token has expired', 401)
+                except InvalidAudience:
+                    raise ApiError('Invalid audience', 401)
+                g.user = jwt.preferred_username
+                g.customers = jwt.customers
+                g.scopes = jwt.scopes
+
+                if not Permission.is_in_scope(scope, g.scopes):
+                    raise ApiError("Missing required scope: %s" % scope, 403)
+                else:
+                    return f(*args, **kwargs)
+
+            # Basic Auth
+            m = re.match(r'Basic (\S+)', auth_header)
+            credentials = m.group(1) if m else None
+
+            if credentials:
+                try:
+                    username, password = base64.b64decode(credentials).decode('utf-8').split(':')
+                except Exception as e:
+                    raise ApiError(str(e), 401)
+
+                user = User.check_credentials(username, password)
+                if not user:
+                    raise ApiError("Invalid username or password", 401)
+                if current_app.config['EMAIL_VERIFICATION'] and not user.email_verified:
+                    raise ApiError('email not verified')
+                if is_authorized('ALLOWED_EMAIL_DOMAINS', groups=[user.domain]):
```

Comment:
The logic of `is_authorized()` seems to be reverse to what you would expect. Check this.